login/tests.py

```diff
@@ -0,0 +1,110 @@
+from django.test import TestCase, Client
+from django.urls import reverse
+from django.contrib.auth import get_user_model
+
+User = get_user_model()
+
+
+class LoginTests(TestCase):
+    def setUp(self):
+        """
+        Set up the test environment, including creating a test user and client.
+        """
+        self.client = Client()
+        self.login_url = reverse('login')
+        self.logout_url = reverse('logout')
+        self.profile_url = reverse('profile')
+
+        # Create a test user
+        self.user = User.objects.create_user(
+            username='testuser',
+            password='password123',
+            email='testuser@example.com',
+            first_name='Test',
+            last_name='User',
+            role='Instructor',
+            phone='1234567890',
+            address='123 Test Street'
+        )
+
+    def test_login_page_get(self):
+        """
+        Test the GET request to load the login page.
+        """
+        response = self.client.get(self.login_url)
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, 'login/login.html')
+        self.assertContains(response, '<h2>Login</h2>')
+
+    def test_login_success(self):
+        """
+        Test successful user login.
+        """
+        response = self.client.post(self.login_url, {
+            'username': 'testuser',
+            'password': 'password123'
+        })
+        self.assertRedirects(response, self.profile_url)
+
+    def test_login_failure(self):
+        """
+        Test user login failure (incorrect password).
+        """
+        response = self.client.post(self.login_url, {
+            'username': 'testuser',
```

Comment:
can we get the fields from self.user instead of having the data written in two separate places?

Reply:
same goes for below cases